### ISSUE TITLE
unscoping (?) #order_as_specified

### ISSUE THREAD

**Issue by taylor-a-barnette** [panorama-ed/order_as_specified#33]
Great gem. However...I am using this gem in the following situation: I am pulling a batch of records and paring the batch down via several class methods that each look to see if the records in the batch meet certain criteria. E.g.: client = Client.all, client.available_in_evenings, client.has_office, client.takes_insurance, etc. Each time the app passes the batch of clients through a filtering method, it will shrink the batch as the app drills down to a list of clients that best fits some user inputs.

This gem comes into play when I pass the clients through a geolocation filter. The order that the filter returns is important, and #order_as_specified preserves that order. Unfortunately, it sets the order and maintains it regardless of how the following methods would order the batch. 

In other words, once #order_as_specified sets the order in a given method, that order persists throughout the rest of the algorithm, all the way down through multiple filtering methods. It doesn't matter if the second or third method produce a different order - the leveraging of #order_as_specified at the outset ensures that all following orderings adhere to whatever order the first method returned.

Is there a way around this? Can this somehow be unscoped? Is there another way to achieve the same output as #order_as_specified?

**Comment by JacobEvelyn** [comment]
Hey @taylor-a-barnette. So `order_as_specified` produces an `ActiveRecord::Relation` object (the same as `Client.all`, `Client.available_in_evenings`, etc.), which means the query won't actually be executed until it needs to. That's the magic built into Rails that allows you to keep calling these methods (often this is called "chaining") without doing a new query each time. For instance, `Client.available_in_evenings.has_office.takes_insurance` only produces one query, instead of hitting your database with three. This is a good thing!

The way `ActiveRecord` does this is by building up on the Ruby side what the SQL query should look like and only executing it when you actually try to do something with the result. `order_as_specified` simply uses the `.order()` method to add an `ORDER BY` SQL condition to the query. You can learn more about how this works [here](http://guides.rubyonrails.org/active_record_querying.html#ordering)—note that in the guide it describes how if you call `.order()` multiple times, the order conditions are _appended together_. So it's not that the additional orderings you're using are ignored, it's just that `order_as_specified`'s ordering takes precedence—so if two things are _tied_ in ordering based on the `order_as_specified` ordering it'll then use the next ordering you add. If you want to _replace_ the order, you can use [the `.reorder()` method](http://guides.rubyonrails.org/active_record_querying.html#reorder) instead.

I'm a little confused about what you're trying to do, however. If you don't want things ordered in the output provided by your `order_as_specified` call, why are you using it there?

**Comment by taylor-a-barnette** [comment]
Thanks @JacobEvelyn for the quick feedback. Vastly appreciated. My point of confusion was centered on what you described in your second paragraph - i.e., what's going on with `.order()`. That said...

Let me clarify what I'm trying to do here. I'm working on an algorithm for a client's site. A previous dev has set-up this algorithm as a Concern so that it behaves as such: 

1. it gets a list of records (from Doctor table)

2. loops that list through 7 class methods to filter down that list and return results to a user based on that user's preferences for a doctor

The relevant code snippet:

```
      search_methods = [
        :match_by_health_type, :match_by_insurance_provider, :match_by_specialties, 
        :match_by_zipcode, :match_by_gender, :match_by_treatment_method, :match_by_availability
      ]
      
      matches = []
      
      search_methods.each.with_index do |method, index|
        prev = query 

        query = query.send(method, user)#.distinct
                
        if query.count(:all) == 0
          query = prev
        elsif query.count(:all) < number_of_matches
          matches << query
          query = prev #set query to previous
        elsif query.count(:all) == number_of_matches
          break
        end
      end
```
I am unsure if this was really the best way to approach things, but I highly doubt the client would be keen on revising this. So...

Here's where I come in. One of the those filters is a method that sorts results by their distance from a user - the `match_by_zipcode`. This is done using a geolocation gem that has a very handy `near` class method that can be used to sort results from closest to farthest. This method acts on the Address model, with which the Doctors model has a has_many relationship. 

#match_by_zipcode should organize this list based on nearest office address to the user.

So, for this geolocation method filter, I added code takes the list of doctors (which has already been filtered through a few other methods) and: 1. gets their addresses; 2. sorts their addresses from closest to farthest; 3. gets an array of the doctor ids of the records that own those addresses, and then pull those records from the Doctor table and orders them according to the addresses using `order_as_specified`.

However, doing this means that I am unable to leverage #distinct (you'll notice it is commented out), as postgres predictably calls that out: `ORDER BY expressions must appear in select list`. Now, the list needs to be distinct with each loop so that if the loop has to break, and results get returned to the user, the user does not get duplicates. 

I thought perhaps there was a way to remove the ordering condition so that distinct can do its job. 

Btw if you have read this far and not just been like, understandably, "not my problem", I applaud you heartily.

**Comment by JacobEvelyn** [comment]
> Btw if you have read this far and not just been like, understandably, "not my problem", I applaud you heartily.

Well now I have to help. 😉 

I suspect you may be making this harder than it needs to be. Are you by chance using the [geocoder](https://github.com/alexreisner/geocoder#for-activerecord-models) gem? (I notice that it provides a `.near` scope...) If so, the `geocoder` docs say:

> by default, objects are ordered by distance.

So I think you should be fine to just use that directly and not use `order_as_specified` in this case.

Does that make sense? Does it help?

**Comment by taylor-a-barnette** [comment]
Lol, thanks, I appreciate it. You know, this issue is not relevant to your gem anymore so at any point feel free to extract yourself - no offense will be taken, you've already done me a tremendous favor by helping up to this point.

RE your suggestion - I believe I did try this and it did not work, but that was several iterations ago and I can't recall for certain. I'll give it another shot and circle back. And you're right that I could be making this far more complicated than it needs to be. Probably been in the weeds on it too much on it.

**Comment by JacobEvelyn** [comment]
No problem; let me know if I can help any more.

**Comment by JacobEvelyn** [comment]
Closing this out, as it seems @taylor-a-barnette's issue was resolved. @taylor-a-barnette, let me know if that's not the case!
